crates/pallet-domains/src/staking_epoch.rs

```diff
@@ -0,0 +1,800 @@
+        }
+
+        operator.current_domain_id = operator.next_domain_id;
+        DomainStakingSummary::<T>::try_mutate(operator.current_domain_id, |maybe_stake_summary| {
+            let stake_summary = maybe_stake_summary
+                .as_mut()
+                .ok_or(TransitionError::MissingDomainStakeSummary)?;
+
+            stake_summary.next_operators.push(operator_id);
+
+            Ok(())
+        })
+    })
+}
+
+fn do_finalize_operator_deregistrations<T: Config>(
+    domain_id: DomainId,
+    consensus_block_number: T::BlockNumber,
+) -> Result<(), Error> {
+    let stake_withdrawal_locking_period = T::StakeWithdrawalLockingPeriod::get();
+    let unlock_block_number = consensus_block_number
+        .checked_add(&stake_withdrawal_locking_period)
+        .ok_or(Error::FinalizeOperatorDeregistration(
+            TransitionError::BlockNumberOverflow,
+        ))?;
+
+    if let Some(operator_ids) = PendingOperatorDeregistrations::<T>::take(domain_id) {
+        PendingUnlocks::<T>::mutate(unlock_block_number, |maybe_stored_operator_ids| {
+            let mut stored_operator_ids = maybe_stored_operator_ids.take().unwrap_or_default();
+            operator_ids.into_iter().for_each(|operator_id| {
+                PendingOperatorUnlocks::<T>::append(operator_id);
+                stored_operator_ids
+                    .retain(|existing_operator_id| *existing_operator_id != operator_id);
+                stored_operator_ids.push(operator_id);
+            });
+            *maybe_stored_operator_ids = Some(stored_operator_ids)
+        })
+    }
+
+    Ok(())
+}
+
+fn unlock_operator<T: Config>(operator_id: OperatorId) -> Result<(), Error> {
+    Operators::<T>::try_mutate(operator_id, |maybe_operator| {
+        let operator = maybe_operator
+            .as_mut()
+            .ok_or(TransitionError::MissingOperator)?;
+
+        let mut total_shares = operator.total_shares;
+        let mut total_stake = operator
+            .current_total_stake
+            .checked_add(&operator.current_epoch_rewards)
+            .ok_or(TransitionError::BalanceOverflow)?;
+
+        let freeze_identifier = T::FreezeIdentifier::staking_freeze_id(operator_id);
+
+        Nominators::<T>::drain_prefix(operator_id).try_for_each(|(nominator_id, nominator)| {
+            let nominator_share = Perbill::from_rational(nominator.shares, total_shares);
+            let nominator_staked_amount = nominator_share.mul_floor(total_stake);
+
+            let locked_amount = T::Currency::balance_frozen(&freeze_identifier, &nominator_id);
+            let amount_to_mint = nominator_staked_amount
+                .checked_sub(&locked_amount)
+                .unwrap_or(Zero::zero());
+
+            // remove the lock and mint any gains
+            T::Currency::thaw(&freeze_identifier, &nominator_id)
+                .map_err(|_| TransitionError::RemoveLock)?;
+            T::Currency::mint_into(&nominator_id, amount_to_mint)
```

Comment:
Nice to check whether amount_to_mint is zero as `mint_into` itself does not seem to do so, a tiny helper function can be helpful.